lib/src/Uncertainty/Process/DiscreteMarkovChain.cxx

```diff
@@ -0,0 +1,243 @@
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "openturns/DiscreteMarkovChain.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/UserDefined.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/Point.hxx"
+#include "openturns/SpecFunc.hxx"
+
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(DiscreteMarkovChain)
+
+static const Factory<DiscreteMarkovChain> Factory_DiscreteMarkovChain;
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain()
+  : ProcessImplementation()
+  , origin_(0)
+  , transitionMatrix_(SquareMatrix(1, 1))
+  , currentState_(0)
+{
+  setOutputDimension(1);
+}
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix)
+  : ProcessImplementation()
+  , origin_(origin)
+  , transitionMatrix_(transitionMatrix)
+  , currentState_(origin)
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+}
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix,
+                       const RegularGrid & timeGrid)
+  : ProcessImplementation()
+  , origin_(origin)
+  , transitionMatrix_(transitionMatrix)
+  , currentState_(origin)
+{
+  setOutputDimension(transitionMatrix.getNbRows());
+  setTimeGrid(timeGrid);
+}
+
+/* Virtual constructor */
+DiscreteMarkovChain * DiscreteMarkovChain::clone() const
+{
+  return new DiscreteMarkovChain(*this);
+}
+
+/* String converter */
+String DiscreteMarkovChain::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << DiscreteMarkovChain::GetClassName()
+      << ", origin=" << origin_
+      << ", transition matrix=" << transitionMatrix_;
+  return oss;
+}
+
+/* String converter */
+String DiscreteMarkovChain::__str__() const
+{
+  OSS oss;
+  oss << getClassName() << "(origin=" << origin_ << ", transition=" << PersistentCollection<Scalar>(*transitionMatrix_.getImplementation()).__str__() << ")";
+  return oss;
+}
+
+/* Realization accessor */
+Field DiscreteMarkovChain::getRealization() const
+{
+  const UnsignedInteger size = mesh_.getVerticesNumber();
+  const UnsignedInteger dimension = transitionMatrix_.getNbRows();
+
+  
+  /* Definition of the states of the system */
+  Sample states(dimension, 1);
+  for (UnsignedInteger i = 0; i < dimension; i++)
+  {
+    states(i, 0) = i;
+  }
+
+  /* Loop on mesh points */
+  Sample data(size, 1);
+  data(0, 0) = origin_;
+  UserDefined dist(states);
+
+  for (UnsignedInteger i = 1; i < size; i++)
+  {
+    /* Definition of the distribution */
+    Point weights(dimension, 0.0);
+    for (UnsignedInteger col = 0; col < dimension; col++)
+    {
+      weights[col] = transitionMatrix_(currentState_, col);
```

Comment:
A column-wise stochastic matrix would have allowed to copy efficiently the probability table of the conditional distribution using 
```
std::copy(transitionMatrix_.begin()+currentState_*dimension, transitionMatrix_.begin()+(1+currentState_)*dimension, weights.begin());
```